enhancements/network/on-prem-service-load-balancers.md

```diff
@@ -0,0 +1,370 @@
+extensive testing.
+
+Given the pre-1.0 beta state of the project, we must pay particular close
+attention to any interfaces that need to be stabilized before we can ship
+MetalLB.  We want to get ahead of potential future upgrade challenges as soon
+as possible.
+
+MetalLB includes two major modes of operation: layer 2 and BGP.  Both have
+strengths and weaknesses.  Multiple modes also means an increase in our test
+matrix.  If this proves to be a challenge, we should consider a phased roll-out
+where we start with only the layer 2 mode (simpler, works in more environments)
+and roll out BGP support as a later stage.
+
+Like any network facing application, MetalLB should be reviewed for any
+security concerns.  This must be part of our ongoing technical due diligence.
+So far, the following areas should receive a close look:
+* The [memberlist](https://github.com/hashicorp/memberlist) protocol and
+  implementation, used by MetalLB's layer 2 mode for cluster membership and
+  fast Node failure detection.
+* MetalLB's [custom implementation of
+  BGP](https://github.com/metallb/metallb/tree/main/internal/bgp)
+* MetalLB's
+  [implementation](https://github.com/metallb/metallb/tree/main/internal/layer2)
+  of [ARP](https://github.com/mdlayher/arp) and
+  [NDP](https://github.com/mdlayher/ndp) for its layer 2 mode.
+
+## Design Details
+
+### Test Plan
+
+Separate test plans are required for the layer 2 and BGP modes of MetalLB.
+
+Testing MetalLB's layer 2 mode will work with our existing `e2e-metal-ipi` job.
+In that CI job, we have full control of the networks used by the installed
+cluster, so we can allocate a range of IP addresses for use by MetalLB.  More
+investigation is needed, but it's likely that we can not test this on our
+`e2e-metal` UPI based jobs because we rely on the network provided by
+packet.net between the cluster hosts.
+
+Testing of the BGP mode is more complicated.  It will require setting up a BGP
+network environment for the cluster nodes to peer with.  We don't have anything
+like this today, so it will take some work.  The upstream MetalLB project needs
+this, as well.  It currently lacks any automated testing of the BGP
+integration.
+
+### Upgrade / Downgrade Strategy
+
+The operator will include any required logic to handle upgrades or downgrades
+and changes between versions of MetalLB.
+
+MetalLB is currently configured via a `ConfigMap` that is not a stable API.  We
+will start by building an operator that provides a stable API for configuration
+and the `ConfigMap` will become an internal implementation detail fully owned
+by the operator.
+
+We must also mitigate these risks through engagement and contributions to the
+upstream community to help make sure that changes made to the software and its
+configuration interfaces can be managed through an upgrade or downgrade
+process.
+
+### Version Skew Strategy
+
+MetalLB has two major components: a single `controller` pod and a `speaker` pod
```

Comment:
does the `speaker` run on every node or a subset?

Reply:
Good question!

I believe "it depends".  In layer 2 mode, it really does need to run on every Node.

In BGP mode, there's more flexibility.  The behavior depends on the `ExternalTrafficPolicy` type on the `Service`.  
- If the type is `cluster`, then `speaker` can run on any subset of Nodes that you want peering with BGP routers.  All of those Nodes will advertise that they can be used to reach all of of the Service load balancer IPs.
- If the type is `local`, then `speaker` must run on every node where an `Endpoint` may exist locally.  Each node will only advertise that load balancer IPs are reachable that can map to a local Endpoint.  Put another way, `speaker` must run on every node where workloads behind a service load balancer may run.

I'll incorporate this information into the enhancement.